code/parse/sexp.cpp

```diff
@@ -17038,511 +17038,280 @@ int sexp_shield_quad_low(int node)
 {
-	int sindex;
-	int requested_bank ;
-	int requested_weapons ;
-	int rearm_limit = -1;
-	bool is_nan, is_nan_forever;
-
-	// Check that a ship has been supplied
-	sindex = ship_name_lookup(CTEXT(node));
-	if (sindex < 0) 
-	{
-		return ;
-	}
-
-	// Get the bank to set the number on
-	requested_bank = eval_num(CDR(node), is_nan, is_nan_forever);
-	if (is_nan || is_nan_forever || requested_bank < 0)
+	int num_banks, *ammo_capacity, *ammo, *start_ammo, *weapons;
+	if (primary)
 	{
-		return ;
+		num_banks = swp->num_primary_banks;
+		ammo_capacity = swp->primary_bank_capacity;
+		ammo = swp->primary_bank_ammo;
+		start_ammo = swp->primary_bank_start_ammo;
+		weapons = swp->primary_bank_weapons;
 	}
-
-	//  Get the number of weapons requested
-	node = CDR(node);
-	requested_weapons = eval_num(CDR(node), is_nan, is_nan_forever);
-	if (is_nan || is_nan_forever || requested_weapons < 0)
+	else
 	{
-		return ;
+		num_banks = swp->num_secondary_banks;
+		ammo_capacity = swp->secondary_bank_capacity;
+		ammo = swp->secondary_bank_ammo;
+		start_ammo = swp->secondary_bank_start_ammo;
+		weapons = swp->secondary_bank_weapons;
 	}
-	
-	node = CDDR(node);	
 
-	// If a rearm limit hasn't been specified simply change the ammo. Otherwise read in the rearm limit
-	if (node >= 0) {
-		rearm_limit = eval_num(node, is_nan, is_nan_forever);
-		if (is_nan || is_nan_forever)
-			return ;
-	}
-	set_primary_ammo(sindex, requested_bank, requested_weapons, rearm_limit);
-
-	// do the multiplayer callback here
-	Current_sexp_network_packet.start_callback();
-	Current_sexp_network_packet.send_ship(sindex);
-	Current_sexp_network_packet.send_int(requested_bank);
-	Current_sexp_network_packet.send_int(requested_weapons);
-	Current_sexp_network_packet.send_int(rearm_limit);
-	Current_sexp_network_packet.end_callback(); 
-}
-
-void multi_sexp_set_primary_ammo()
-{
-	int sindex;
-	int requested_bank;
-	int requested_weapons;
-	int rearm_limit;
-
-	Current_sexp_network_packet.get_ship(sindex);
-	Current_sexp_network_packet.get_int(requested_bank);
-	Current_sexp_network_packet.get_int(requested_weapons);
-	Current_sexp_network_packet.get_int(rearm_limit);
-
-	set_primary_ammo(sindex, requested_bank, requested_weapons, rearm_limit);
-}
-
-//Karajorma - Helper function for the set-primary-ammo and weapon functions
-void set_primary_ammo (int ship_index, int requested_bank, int requested_ammo, int rearm_limit)
-{
-	ship *shipp;
-	int maximum_allowed ;
-
-	// Check that it's valid
-	if((Ships[ship_index].objnum < 0) || (Ships[ship_index].objnum >= MAX_OBJECTS)){
-		return ;
-	}
-	shipp = &Ships[ship_index];
-	
 	// Can only set one bank at a time. Check that someone hasn't asked for the contents of all 
 	// the banks or a non-existant bank
-	if ((requested_bank > shipp->weapons.num_primary_banks) || requested_bank < 0)
-	{
-		return ;
-	}
+	if ((requested_bank > num_banks) || requested_bank < 0)
+		return;
```

Comment:
The old version didn't have this but this case would indicate that there is something wrong in the mission, right? If so, then it would be nice for the mission designer to have a warning message here to explain why nothing is happening.

Same applies to the ballistics check below.

Reply:
That actually points to a legit issue; it should be `>= num_banks`. :)

Unfortunately a warning message isn't really possible here from a conceptual standpoint.  Getting ammo is more subtle and a warning would be more appropriate there; but the most important edge case is also a valid use of that sexp.  A bank number on one ship can be the number of banks on another ship, and there's no way to tell what the mission designer intended.  So putting a warning on `get` won't work, and putting a warning only on `set` would result in a false sense of security on `get`.

Reply:
Uhh, yes. I totally noticed that!

Yeah, those are good reasons for not including a warning. Was just a suggestion anyway.